File: js/dist-test/src/protocol.js
/**
 * Wire envelopes — same JSON shape as Python's zhub.protocol.
 */
import { randomUUID } from 'crypto';
export function newRequestId() {
    // Match Python's uuid4().hex shape: 32 lowercase hex chars (no dashes).
    return randomUUID().replace(/-/g, '');
}
export function envelope(type, payload = {}, requestId) {
    return { type, request_id: requestId ?? newRequestId(), payload };
}
export function registerPublisher(manifest, desiredName, apiKey) {
    const payload = { manifest };
    if (desiredName !== undefined)
        payload.desired_name = desiredName;
    if (apiKey !== undefined)
        payload.api_key = apiKey;
    return envelope('register-publisher', payload);
}
export function registerConnection(aiName, apiKey, clientManifest) {
    return envelope('register-connection', {
        ai_name: aiName,
        api_key: apiKey,
        client_manifest: clientManifest,
    });
}
export function chatRequest(messages, model = 'default', temperature = 0.4, maxTokens = 4096, extras) {
    const payload = {
        messages,
        model,
        temperature,
        max_tokens: maxTokens,
    };
    if (extras)
        Object.assign(payload, extras);
    return envelope('chat-request', payload);
}
export function chatResponse(text, requestId, finishReason = 'stop') {
    return envelope('chat-response', { text, finish_reason: finishReason, tool_calls: [], usage: {} }, requestId);
}
export function chatChunk(delta, requestId, done = false, finishReason) {
    return envelope('chat-chunk', { delta, done, finish_reason: finishReason ?? null }, requestId);
}
export function invokeRequest(connectionId, capability, args) {
    return envelope('invoke-request', {
        connection_id: connectionId,
        capability,
        args,
    });
}
export function invokeResult(requestId, ok, result, error) {
    return envelope('invoke-result', { ok, result: result ?? null, error: error ?? null }, requestId);
}
export function errorEnvelope(requestId, code, message) {
    return envelope('error', { code, message }, requestId);
}

File: kotlin/src/main/kotlin/com/zawwar/zhub/Protocol.kt
package com.zawwar.zhub

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.JsonPrimitive
import java.util.UUID

/**
 * Wire envelope. Matches the Python implementation byte-for-byte.
 *
 * type:        message kind discriminator
 * request_id:  correlates request/response pairs
 * payload:     type-specific contents
 */
@Serializable
data class Envelope(
    val type: String,
    val request_id: String = newRequestId(),
    val payload: JsonObject = JsonObject(emptyMap()),
) {
    fun toJson(): String = json.encodeToString(this)

    companion object {
        fun fromJson(text: String): Envelope = json.decodeFromString(text)
    }
}

internal val json = Json {
    ignoreUnknownKeys = true
    encodeDefaults = false
}

internal fun newRequestId(): String = UUID.randomUUID().toString().replace("-", "")

// ---- helper constructors ------------------------------------------------

internal fun registerConnectionEnvelope(
    aiName: String,
    apiKey: String,
    clientManifest: Manifest,
): Envelope = Envelope(
    type = "register-connection",
    payload = buildJsonObject {
        put("ai_name", JsonPrimitive(aiName))
        put("api_key", JsonPrimitive(apiKey))
        put("client_manifest", json.encodeToJsonElement(Manifest.serializer(), clientManifest) as JsonObject)
    },
)

internal fun chatRequestEnvelope(
    messages: List<Map<String, String>>,
    model: String = "default",
    temperature: Double = 0.4,
    maxTokens: Int = 4096,
): Envelope {
    // Build the messages array explicitly — generic List<Map<...>> doesn't
    // resolve cleanly through Json.encodeToJsonElement on Kotlin 2.0+, and an
    // explicit construction avoids the reified-type ambiguity entirely.
    val messagesArray = kotlinx.serialization.json.JsonArray(
        messages.map { m ->
            buildJsonObject {
                for ((k, v) in m) put(k, JsonPrimitive(v))
            }
        }
    )
    return Envelope(
        type = "chat-request",
        payload = buildJsonObject {
            put("messages", messagesArray)
            put("model", JsonPrimitive(model))
            put("temperature", JsonPrimitive(temperature))
            put("max_tokens", JsonPrimitive(maxTokens))
        },
    )
}

internal fun invokeResultEnvelope(
    requestId: String,
    ok: Boolean,
    result: JsonElement? = null,
    error: String? = null,
): Envelope = Envelope(
    type = "invoke-result",
    request_id = requestId,
    payload = buildJsonObject {
        put("ok", JsonPrimitive(ok))
        if (result != null) put("result", result)
        if (error != null) put("error", JsonPrimitive(error))
    },
)
